vsphere/resource_vsphere_vapp.go

```diff
@@ -0,0 +1,723 @@
+							Optional: true,
+						},
+					},
+				},
+			},
+			"template_vapp": &schema.Schema{
+				Type:     schema.TypeList,
+				Optional: true,
+				MaxItems: 1,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"name": &schema.Schema{
+							Type:     schema.TypeString,
+							Required: true,
+						},
+						"disk_provisioning": &schema.Schema{
+							Type:     schema.TypeString,
+							Optional: true,
+							Default:  types.VAppCloneSpecProvisioningTypeSameAsSource,
+						},
+						"network_mapping": &schema.Schema{
+							Type:     schema.TypeSet,
+							Optional: true,
+							Elem: &schema.Resource{
+								Schema: map[string]*schema.Schema{
+									"source_network_label": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+									},
+									"destination_network_label": &schema.Schema{
+										Type:     schema.TypeString,
+										Required: true,
+									},
+								},
+							},
+						},
+					},
+				},
+			},
+		},
+	}
+}
+
+func resourceVSphereVAppCreate(d *schema.ResourceData, meta interface{}) error {
+	client := meta.(*govmomi.Client)
+
+	log.Printf("[DEBUG] resourceVSphereVAppCreate :: client : %#v", client)
+	log.Printf("[DEBUG] resourceVSphereVAppCreate :: ResourceData d: %#v", d)
+
+	// Construct vAPP Object with required Attributes
+	vapp := vApp{
+		d:    d,
+		c:    client,
+		name: d.Get("name").(string),
+	}
+
+	err := vapp.populateOptionalVAppAttributes(d)
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	err = vapp.populateVAppTemplate(d)
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	err = vapp.populateVAppEntities(d)
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	err = vapp.populateVAppResourceAllocationInfo(d)
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	log.Printf("[DEBUG] resourceVSphereVAppCreate :: vapp : %#v", vapp)
+
+	// VApp Creation
+
+	err = vapp.calculateLocation()
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	err = vapp.create()
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+
+	err = vapp.linkEntities()
+	if err != nil {
+		log.Printf("[ERROR] %s", err)
+		return err
+	}
+	d.SetId(vapp.name)
+	return resourceVSphereVAppRead(d, meta)
+}
+
+func resourceVSphereVAppRead(d *schema.ResourceData, meta interface{}) error {
+
+	log.Printf("[DEBUG] resourceVSphereVAppRead:: d : %#v", d)
+
+	d.Set("name", d.Get("name"))
+
+	log.Printf("[DEBUG] resourceVSphereVAppRead:: d : %#v", d)
+
+	return nil
+}
+func resourceVSphereVAppUpdate(d *schema.ResourceData, meta interface{}) error {
+	return nil
+}
+
+func resourceVSphereVAppDelete(d *schema.ResourceData, meta interface{}) error {
+
+	client := meta.(*govmomi.Client)
+
+	log.Printf("[DEBUG] resourceVSphereVAppDelete:: ResourceData d: %#v", d)
+
+	dc, err := getDatacenter(client, d.Get("datacenter").(string))
```

Comment:
This function is used in both delete and create operation. Hence only common things are placed under getCreatedVApp( ) function.